library/core/src/mem/mod.rs

```diff
@@ -909,11 +912,12 @@ pub fn take<T: Default>(dest: &mut T) -> T {
 #[rustc_const_unstable(feature = "const_replace", issue = "83164")]
 #[cfg_attr(not(test), rustc_diagnostic_item = "mem_replace")]
 pub const fn replace<T>(dest: &mut T, src: T) -> T {
+    let dest: *mut T = dest;
     // SAFETY: We read from `dest` but directly write `src` into it afterwards,
     // such that the old value is not duplicated. Nothing is dropped and
     // nothing here can panic.
     unsafe {
-        let result = ptr::read(dest);
+        let result = ptr::read_mut(dest);
```

Comment:
Compared to `ptr::read`, this just avoids a single MIR statement to turn the `*mut` into a `*const`, right? I am quite surprised that this would be a change worth doing. And if it is worth doing, a more holistic approach might be to just not do that cast and instead allow `*mut` everywhere that `*const` is allowed?

Reply:
This does that a bit more for the *method* versions -- `*mut` has had `read` [for 5 years now](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1260-2018-05-10), and this PR makes that 1 statement shorter, which is certainly unimportant on its own but impacts every time it's mir-inlined.  I suppose we can compatibly add `impl Trait` parameters to functions now, so we could consider letting `ptr::read` take more things too, but I don't know how disruptive that would be -- or if libs-api would even want it.

But I'm also not certain I understand what you were imagining with "allow `*mut` everywhere that `*const` is allowed".  Could you sketch it out a bit more?  I'm curious.  (AFAIK MIR is typed enough that we can't omit these casts, even though they're NOPs?)

Reply:
I was thinking MIR optimizations could remove the `*mut` to `*const` cast that was introduced by MIR building, and our notion of "well-typed MIR" could be weakened to say that this is fine.